crates/red_knot_python_semantic/resources/mdtest/narrow/truthiness.md

```diff
@@ -219,3 +219,49 @@ else:
 # TODO: It should be A. We should improve UnionBuilder or IntersectionBuilder. (issue #15023)
 reveal_type(y)  # revealed: A & ~AlwaysTruthy | A & ~AlwaysFalsy
 ```
+
+## Filtering in chained boolean expressions
+
+```py
+from typing import Literal
+
+class A: ...
+
+def _(x: Literal[0, 1]):
+    reveal_type(x or A())  # revealed: Literal[1] | A
+    reveal_type(x and A())  # revealed: Literal[0] | A
+
+# Filtering should not happen in cases where it makes the result type more complex
+def _(x: str):
+    reveal_type(x or A())  # revealed: str | A
+    reveal_type(x and A())  # revealed: str | A
```

Comment:
This is not part of the requirements in the linked task. For the scope of this task, we should simply intersect with `~AlwaysTruthy` or `~AlwaysFalsy` and let the resulting types shake out.

If we feel we need to simplify display of some types to avoid user confusion, we may do that in future, but we want to evaluate that holistically in the context of real-world code, not pre-emptively. And we will implement that in a general way (either in type display or in intersection building), not special-cased to a single narrowing scenario.
```suggestion
def _(x: str):
    reveal_type(x or A())  # revealed: str & ~AlwaysFalsy | A
    reveal_type(x and A())  # revealed: str & ~AlwaysTruthy | A
```